File: android/app/src/main/kotlin/app/zen/chromium/SiteData.kt
package app.zen.chromium

import android.os.Handler
import android.os.Looper
import android.webkit.CookieManager
import android.webkit.WebStorage
import androidx.webkit.WebStorageCompat
import androidx.webkit.WebViewFeature
import app.zen.chromium.blocking.Domains
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.Executors

/**
 * The WebView's knowledge of one site for the site-information sheet and of every site for the
 * site-data viewer: which cookies a page receives, how much a site stores, and how to take both
 * away again – per container, since every container is its own WebView profile with its own jar
 * and quota manager. Only names and attributes leave this class; cookie values stay in the jar.
 * The reasoning about the jar's answers is in [CookieJar], which has no Android in it.
 */
class SiteData {
    private val main = Handler(Looper.getMainLooper())

    /** The viewer's jar reads (one `getCookie` per origin, hundreds of them) run here, off the main thread. */
    private val reader = Executors.newSingleThreadExecutor { r -> Thread(r, "zen-sitedata") }

    /** `[{ name, domain, secure, size }]` for the page at `url`. */
    fun cookies(containerId: String, url: String): JSONArray {
        val target = CookieJar.Target.of(url) ?: return JSONArray()
        val jar = Profiles.cookieManager(containerId)
        return CookieJar.toJson(CookieJar.classify(target) { CookieJar.parse(jar.getCookie(it)) })
    }

    /**
     * Expire every cookie the page at `url` receives, under each domain and path it may be
     * scoped to, then read again: `{ removed, remaining }`.
     */
    fun clearCookies(containerId: String, url: String, reply: (Any?) -> Unit) {
        val target = CookieJar.Target.of(url)
        if (target == null) {
            reply(json("removed" to 0, "remaining" to 0))
            return
        }
        val jar = Profiles.cookieManager(containerId)
        val before = CookieJar.parse(jar.getCookie(target.pageUrl))
        val headers = CookieJar.expiryHeaders(before.map { it.name }, target)
        if (headers.isEmpty()) {
            reply(json("removed" to 0, "remaining" to before.size))
            return
        }
        var pending = headers.size
        val finish = {
            jar.flush()
            val remaining = CookieJar.parse(jar.getCookie(target.pageUrl)).size
            reply(json("removed" to (before.size - remaining).coerceAtLeast(0), "remaining" to remaining))
        }
        for ((setUrl, header) in headers) {
            jar.setCookie(setUrl, header) { main.post { if (--pending == 0) finish() } }
        }
    }

    /**
     * Quota-managed storage of `site` (IndexedDB, Cache API, …): the origins of the site that
     * hold data and their summed usage. Web Storage is not quota-managed on Android; the page
     * itself reports that part.
     */
    fun storage(containerId: String, site: String, reply: (Any?) -> Unit) {
        val storage = Profiles.webStorage(containerId)
        storage.getOrigins { raw ->
            val origins = JSONArray()
            var usage = 0L
            var quota = 0L
            val entries = (raw as? Map<*, *>)?.values ?: emptyList<Any?>()
            for (entry in entries) {
                val origin = entry as? WebStorage.Origin ?: continue
                if (!CookieJar.originBelongsTo(origin.origin, site)) continue
                origins.put(origin.origin)
                usage += origin.usage
                quota = maxOf(quota, origin.quota)
            }
            reply(
                json(
                    "usageBytes" to usage,
                    "quotaBytes" to if (quota > 0) quota else JSONObject.NULL,
                    "origins" to origins
                )
            )
        }
    }

    /**
     * Every origin of the container with data, for the site-data viewer: the quota manager's
     * origins with their usage (`WebStorage.getOrigins`) and, of the `probe` origins the core
     * knows of (visited, with permissions), those holding cookies – WebView's jar cannot be
     * enumerated, so an origin that left cookies alone is found only when the core asks after
     * it. A cookie count is what a page of the origin receives (`getCookie`), so a domain cookie
     * counts under every host of its site the list holds. `[{ origin, cookies, usageBytes }]`,
     * at most [ORIGIN_LIMIT] rows; the jar is read off the main thread.
     */
    fun listOrigins(containerId: String, probe: JSONArray, reply: (Any?) -> Unit) {
        val jar = Profiles.cookieManager(containerId)
        val storage = Profiles.webStorage(containerId)
        storage.getOrigins { raw ->
            val stored = ArrayList<Pair<String, Long>>()
            val entries = (raw as? Map<*, *>)?.values ?: emptyList<Any?>()
            for (entry in entries) {
                val origin = entry as? WebStorage.Origin ?: continue
                val key = Domains.originOf(origin.origin) ?: continue
                stored.add(key to origin.usage)
                if (stored.size >= ORIGIN_LIMIT) break
            }
            val probes = ArrayList<String>()
            for (i in 0 until probe.length()) {
                if (probes.size >= PROBE_LIMIT) break
                Domains.originOf(probe.optString(i))?.let { if (it !in probes) probes.add(it) }
            }
            reader.execute {
                val rows = originRows(stored, probes) { origin -> CookieJar.parse(runCatching { jar.getCookie("$origin/") }.getOrNull()).size }
                main.post { reply(rows) }
            }
        }
    }

    /**
     * Delete what `site` stored. WebViews that can (`DELETE_BROWSING_DATA`) wipe the whole site –
     * cookies, Web Storage, IndexedDB, caches, service workers, across its subdomains – in one
     * call; older ones delete the quota-managed data of the given origins (the page has already
     * cleared its own Web Storage and unregistered its workers by then).
     */
    fun clearStorage(containerId: String, site: String, origins: JSONArray, reply: (Any?) -> Unit) {
        val storage = Profiles.webStorage(containerId)
        if (site.isNotEmpty() && WebViewFeature.isFeatureSupported(WebViewFeature.DELETE_BROWSING_DATA)) {
            val started = runCatching {
                WebStorageCompat.deleteBrowsingDataForSite(storage, site) { main.post { reply(json("ok" to true, "scope" to "site")) } }
            }
            if (started.isSuccess) return
        }
        for (i in 0 until origins.length()) {
            val origin = origins.optString(i)
            if (origin.isNotEmpty()) runCatching { storage.deleteOrigin(origin) }
        }
        reply(json("ok" to true, "scope" to "origins"))
    }

    companion object {
        /** The viewer's cap (`SITE_DATA_ORIGIN_CAP` in the core). */
        const val ORIGIN_LIMIT = 1000

        /** Probed origins per call: each is one synchronous read of the jar. */
        const val PROBE_LIMIT = 500

        /**
         * The viewer's rows from the quota manager's `stored` origins (with their usage) and the
         * core's `probes`, each origin's cookies counted through `cookies`: a stored origin is a
         * row whatever it holds; a probed one only when it holds cookies. Pure, for the tests.
         */
        fun originRows(stored: List<Pair<String, Long>>, probes: List<String>, cookies: (String) -> Int): JSONArray {
            val out = JSONArray()
            val seen = HashSet<String>()
            for ((origin, usage) in stored) {
                if (!seen.add(origin)) continue
                out.put(json("origin" to origin, "cookies" to cookies(origin), "usageBytes" to usage))
            }
            for (origin in probes) {
                if (out.length() >= ORIGIN_LIMIT) break
                if (!seen.add(origin)) continue
                val count = cookies(origin)
                if (count > 0) out.put(json("origin" to origin, "cookies" to count, "usageBytes" to JSONObject.NULL))
            }
            return out
        }
    }
}

File: src/android/siteData.ts
import type { SiteCookie } from '@shared/siteInfo'
import type { SiteDataHost, SiteDataOriginReading, SiteStorageReading } from '@core/platform'
import type { Bridge } from './bridge'

/** What Kotlin's `site.cookies` answers for one cookie (see `CookieJar.Classified`). */
export interface NativeCookie {
  name: string
  domain: string
  secure: boolean | null
  size: number
}

/** What Kotlin's `site.listOrigins` answers for one origin (`SiteData.listOrigins`). */
export interface NativeOriginReading {
  origin: string
  cookies: number
  usageBytes: number | null
}

/**
 * Cookies and storage per site, read from the container's WebView profile. Kotlin can only infer
 * cookie attributes from the jar's answers for related URLs, so path and expiry stay unknown and
 * HttpOnly is left for the core to work out from the page's own `document.cookie`. WebView's
 * jar cannot be enumerated either: the site-data viewer's origins are the quota manager's
 * (`WebStorage.getOrigins`, with their usage) and the ones the core probes for cookies.
 */
export class AndroidSiteData implements SiteDataHost {
  constructor(private readonly bridge: Bridge) {}

  async listOrigins(containerId: string, probe: string[]): Promise<SiteDataOriginReading[]> {
    const raw = await this.bridge.call<NativeOriginReading[] | null>('site.listOrigins', {
      containerId,
      probe
    })
    return (Array.isArray(raw) ? raw : []).flatMap((entry) => {
      const reading = originReadingFromNative(entry)
      return reading ? [reading] : []
    })
  }

  async cookies(containerId: string, url: string): Promise<SiteCookie[]> {
    const raw = await this.bridge.call<NativeCookie[] | null>('site.cookies', { containerId, url })
    return (Array.isArray(raw) ? raw : []).map(cookieFromNative)
  }

  async storage(containerId: string, site: string): Promise<SiteStorageReading> {
    const raw = await this.bridge.call<{
      usageBytes?: number | null
      quotaBytes?: number | null
      origins?: string[]
    } | null>('site.storage', { containerId, site })
    return {
      usageBytes: typeof raw?.usageBytes === 'number' ? raw.usageBytes : null,
      quotaBytes: typeof raw?.quotaBytes === 'number' ? raw.quotaBytes : null,
      origins: Array.isArray(raw?.origins) ? raw.origins.map(String) : []
    }
  }

  async clearCookies(containerId: string, url: string): Promise<number> {
    const raw = await this.bridge.call<{ removed?: number } | null>('site.clearCookies', {
      containerId,
      url
    })
    return typeof raw?.removed === 'number' ? raw.removed : 0
  }

  async clearStorage(containerId: string, site: string, origins: string[]): Promise<void> {
    await this.bridge.call('site.clearStorage', { containerId, site, origins })
  }
}

/** One origin row as the core takes it; null for an entry without an origin. */
export function originReadingFromNative(raw: Partial<NativeOriginReading> | null): SiteDataOriginReading | null {
  if (!raw || typeof raw.origin !== 'string' || !raw.origin) return null
  const cookies = typeof raw.cookies === 'number' && Number.isFinite(raw.cookies) ? Math.max(0, raw.cookies) : 0
  const usage =
    typeof raw.usageBytes === 'number' && Number.isFinite(raw.usageBytes) ? Math.max(0, raw.usageBytes) : null
  return { origin: raw.origin, cookies, usageBytes: usage }
}

export function cookieFromNative(raw: NativeCookie): SiteCookie {
  return {
    name: typeof raw.name === 'string' ? raw.name : '',
    domain: typeof raw.domain === 'string' ? raw.domain : '',
    path: '',
    secure: typeof raw.secure === 'boolean' ? raw.secure : null,
    httpOnly: null,
    session: null,
    size: typeof raw.size === 'number' && Number.isFinite(raw.size) ? Math.max(0, raw.size) : 0
  }
}
